File: src/services/watchlist-service.ts
//services/watchlist-service.ts

import pool from '../config/database';

export interface IWatchlistItem {
    id: number;
    stock_id: number;
    symbol: string;
    name: string;
    created_at: Date;
}

export class WatchlistService {
    // Ambil semua watchlist user
    static async getWatchlist(userId: string): Promise<IWatchlistItem[]> {
        const result = await pool.query(`
            /* dialect: postgres */
            SELECT 
                w.id,
                w.stock_id,
                s.symbol,
                s.name,
                w.created_at
            FROM watchlists w
            JOIN stocks s ON w.stock_id = s.id
            WHERE w.user_id = $1
            ORDER BY w.created_at DESC
        `, [userId]);

        return result.rows;
    }

    // Tambah saham ke watchlist
    static async addToWatchlist(userId: string, symbol: string): Promise<IWatchlistItem> {
        // Cari stock_id berdasarkan symbol
        const stockRes = await pool.query(
            'SELECT id, symbol, name FROM stocks WHERE symbol = $1 AND is_active = true',
            [symbol.toUpperCase()]
        );

        if ((stockRes.rowCount ?? 0) === 0) {
            throw new Error('Saham tidak ditemukan atau tidak aktif');
        }

        const stock = stockRes.rows[0];

        // Insert ke watchlist
        const result = await pool.query(`
            /* dialect: postgres */
            INSERT INTO watchlists (user_id, stock_id)
            VALUES ($1, $2)
            ON CONFLICT (user_id, stock_id) DO NOTHING
            RETURNING id, stock_id, created_at
        `, [userId, stock.id]);

        if ((result.rowCount ?? 0) === 0) {
            throw new Error('Saham sudah ada di watchlist');
        }

        return {
            id: result.rows[0].id,
            stock_id: stock.id,
            symbol: stock.symbol,
            name: stock.name,
            created_at: result.rows[0].created_at
        };
    }

    // Hapus saham dari watchlist
    static async removeFromWatchlist(userId: string, symbol: string): Promise<boolean> {
        // Cari stock_id berdasarkan symbol
        const stockRes = await pool.query(
            'SELECT id FROM stocks WHERE symbol = $1',
            [symbol.toUpperCase()]
        );

        if ((stockRes.rowCount ?? 0) === 0) {
            throw new Error('Saham tidak ditemukan');
        }

        const result = await pool.query(`
            /* dialect: postgres */
            DELETE FROM watchlists
            WHERE user_id = $1 AND stock_id = $2
            RETURNING id
        `, [userId, stockRes.rows[0].id]);

        if ((result.rowCount ?? 0) === 0) {
            throw new Error('Saham tidak ada di watchlist');
        }

        return true;
    }

    // Cek apakah saham ada di watchlist user
    static async isInWatchlist(userId: string, symbol: string): Promise<boolean> {
        const stockRes = await pool.query(
            'SELECT id FROM stocks WHERE symbol = $1',
            [symbol.toUpperCase()]
        );

        if ((stockRes.rowCount ?? 0) === 0) {
            return false;
        }

        const result = await pool.query(`
            /* dialect: postgres */
            SELECT 1 FROM watchlists
            WHERE user_id = $1 AND stock_id = $2
        `, [userId, stockRes.rows[0].id]);

        return (result.rowCount ?? 0) > 0;
    }
}

File: go-backend/services/watchlist_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"mbit-backend-go/config"

	"github.com/jackc/pgx/v5"
)

type WatchlistItem struct {
	ID        int       `json:"id"`
	StockID   int       `json:"stock_id"`
	Symbol    string    `json:"symbol"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at"`
}

type WatchlistService struct{}

var GlobalWatchlistService = &WatchlistService{}

func (s *WatchlistService) GetWatchlist(userID string) ([]WatchlistItem, error) {
	ctx := context.Background()
	query := `
		SELECT
			w.id,
			w.stock_id,
			s.symbol,
			s.name,
			w.created_at
		FROM watchlists w
		JOIN stocks s ON w.stock_id = s.id
		WHERE w.user_id = $1
		ORDER BY w.created_at DESC
	`
	rows, err := config.DB.Query(ctx, query, userID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var items []WatchlistItem
	for rows.Next() {
		var item WatchlistItem
		if err := rows.Scan(&item.ID, &item.StockID, &item.Symbol, &item.Name, &item.CreatedAt); err != nil {
			continue
		}
		items = append(items, item)
	}

	if items == nil {
		items = []WatchlistItem{}
	}

	return items, nil
}

func (s *WatchlistService) AddToWatchlist(userID string, symbol string) (*WatchlistItem, error) {
	ctx := context.Background()

	// Check stock
	var stockID int
	var stockName string
	err := config.DB.QueryRow(ctx, "SELECT id, name FROM stocks WHERE symbol = $1 AND is_active = true", symbol).Scan(&stockID, &stockName)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, fmt.Errorf("saham tidak ditemukan atau tidak aktif")
		}
		return nil, err
	}

	// Insert
	var item WatchlistItem
	err = config.DB.QueryRow(ctx, `
		INSERT INTO watchlists (user_id, stock_id)
		VALUES ($1, $2)
		ON CONFLICT (user_id, stock_id) DO NOTHING
		RETURNING id, stock_id, created_at
	`, userID, stockID).Scan(&item.ID, &item.StockID, &item.CreatedAt)

	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, fmt.Errorf("saham sudah ada di watchlist")
		}
		return nil, err
	}

	item.Symbol = symbol
	item.Name = stockName

	return &item, nil
}

func (s *WatchlistService) RemoveFromWatchlist(userID string, symbol string) error {
	ctx := context.Background()

	// Get stock id
	var stockID int
	err := config.DB.QueryRow(ctx, "SELECT id FROM stocks WHERE symbol = $1", symbol).Scan(&stockID)
	if err != nil {
		if err == pgx.ErrNoRows {
			return fmt.Errorf("saham tidak ditemukan")
		}
		return err
	}

	// Delete
	tag, err := config.DB.Exec(ctx, "DELETE FROM watchlists WHERE user_id = $1 AND stock_id = $2", userID, stockID)
	if err != nil {
		return err
	}

	if tag.RowsAffected() == 0 {
		return fmt.Errorf("saham tidak ada di watchlist")
	}

	return nil
}
